### PR TITLE
feat(content, port): port treetops and ability to climb trees from DDA

### PR DESCRIPTION
**Pull request by ethanmeade**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [ ] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [ ] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

## Purpose of change

Ports over the ability to climb trees from DDA as implemented [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/70669). This does not port over the mutation changes regarding that PR; it also does not include tiles for the treetops themselves. 

## Describe the solution

Cherry-picked the code and JSON changes from the DDA PR to this code-base. This includes the addition of four new "treetop" terrains to the terrain-roofs JSON file and modifications of the tree entries in the terrain-flora JSON file. It also implements two new terrain flags: SINGLE_SUPPORT, which causes the terrain to be destroyed if the support under it is removed, even if there are adjacent supports, and CLIMB_ADJACENT, which allows the player to climb the terrain while standing adjacent to it. 

## Describe alternatives you've considered

Rewriting the code from scratch instead of porting it over from DDA. Alternatively, implementing no changes. 

## Testing

- [x] Generate a world with new trees
- [x] Climb tree without throwing an exception / generating errors
- [x] Climb down from tree without generating errors
- [ ] Fall off of tree
- [x] Cut down a tree and check that the treetop is destroyed with it
  - [x] Cut down a tree with supporting walls nearby and confirm that treetop is still destroyed

### CODE REVIEW

**Comment by ethanmeade** [comment]
Currently, the game generates the world but throws errors while doing so. Actually attempting to climb a tree results in a crash. Working on debugging it now.

File: src/game.cpp [hunk 1]
```diff
@@ -10031,7 +10039,7 @@ void game::vertical_move( int movez, bool force, bool peeking )
 
         const auto cost = map_funcs::climbing_cost( m, u.pos(), stairs );
 
-        if( !cost.has_value() ) {
+        if( !cost.has_value() && !adjacent_climb ) {
```
Comment (scarf005) [inline src/game.cpp:10042]: ### issue

this part crashes the vertical move code.

1. `cost` is of type `const std::optional<int>`, and attempting to get value when it's empty throws exception. you need to check it does has value using `cost.has_value` below.
2. originally, it made an early return if `cost` was empty in the below code block.
3. however, in the new condition, the code block will not get executed when `cost.has_value() == false` and `adjacent_climb == true`
4. which means it skips `cost` validation in https://github.com/ethanmeade/cataclysm-bn-local/blob/25b13fdb0aac0cb6d3a9ffc48ea6509198ab2acc/src/game.cpp#L10077 , hence the crash

### possible solution

make sure `cost` is validated before getting evaluated.

**Review by scarf005 — CHANGES_REQUESTED** [review]

**Comment by ethanmeade** [comment]
Changed the trees to use the current CLIMBABLE terrain flag rather than the ported-over CLIMBABLE_ADJACENT terrain flag. Now it no longer crashes! Some issues do remain though: explicitly cutting a tree down does not remove the treetop, and when the treetop is removed it leaves behind an ASCII air tile rather than the normal graphical tile. 

I'm not sure how difficult the trees should be to climb / move between, but I'm open to discussions on how that should be adjusted.

**Comment by ethanmeade** [comment]
So now treetops collapse whenever their parent tree is cutdown, but adjacent treetops also seem to collapse upon this happening, even if their parent trees are out of the way of the felled tree. 

Amusingly enough, this behavior appears to also occur in DDA. Nevertheless, it doesn't seem that desirable, so I'm looking into preventing it.

**Comment by KheirFerrum** [comment]
Quick note here, `SINGLE_SUPPORT` flag is unnecessary, all behaviour is functionally identical with and without it. Roof's will still collapse with changes made, and thus far I have been unable to confirm the source of said behaviour except that it has nothing to do with `collapse_check()`, `collapse_at()` or `propagate_suspension_check`

Please remove `SINGLE_SUPPORT` in light of this to prevent code from becoming more complex for no benefit.

**Review by KheirFerrum — CHANGES_REQUESTED** [review]
`SINGLE_SUPPORT` flag is unnecessary, all behaviour is functionally identical with and without it. Roof's will still randomly collapse with changes made. Previous commit I did has removed the ability for roofs to collapse if they are supported by walls underneath. It also explains the edge case with this that wouldn't be resolved by `SINGLE_SUPPORT` as it is now but also doesn't actually need to be "fixed" in my opinion.

Please remove `SINGLE_SUPPORT` in light of this to prevent code from becoming more complex for no benefit.

**Review by KheirFerrum — APPROVED** [review]
Made a number of changes to the code and json. Should be ready for merging.
There's an unrelated issue where collapsed rooves form open air that displays as a placeholder instead of being transparent, but that bug is out of scope for this PR.

**Review by scarf005 — APPROVED** [review]
@chaosvolt could you playtest this once more?

**Comment by ethanmeade** [comment]
Sorry about not contributing further, got sidetracked and forgot about this. Let me know if there's anything I can do to help - although it looks like its already been handled outside of the play testing

**Review by RoyalFox2140 — APPROVED** [review]
Compiled. Load Tested. 

Climbing lone and multiple treetops works, smashing wood chopping and trees colliding into each other on wood chopping all works as intended.

Roof collapse on wall and floor destruction works as it used to even though this still means its inconsistent and allows for islands. Everything works and roof collapse sucking is out of scope.

**Comment by KheirFerrum** [comment]
> Sorry about not contributing further, got sidetracked and forgot about this. Let me know if there's anything I can do to help - although it looks like its already been handled outside of the play testing

No issue, it's mostly done now, not a lot of the C++ side remains from DDA's implementation, and the JSON side has seen extensive modification, but hell, it's not the first time we've had PRs end up being loosely inspired.

**Pull request merged**